### PR TITLE
[dashboard] implement feedback modal 9928

### LINKED ISSUE #9928: Implement feedback modal
154722195-8fbfc38e-df99-492e-aade-83c6c2c38116

### PR DESCRIPTION
**Pull request by laushinka**
## Description
Implements feedback tab on the top nav and shows a feedback modal on click.
This also sends a tracking event to Segment with the event name `feedback_submitted`.

| First flow | Second flow | Analytics |
| - | - | - |
| Screenshot 2022-05-18 at 17 42 45 | Screenshot 2022-05-18 at 17 42 55 | Screenshot 2022-05-17 at 15 17 34 |

## Related Issue(s)

Fixes #9928 

#### Functionality test:
1. Go to the top nav and click on the `Feedback` tab. (Ideally, this change in the nav would be hidden behind a feature flag if [this PR](https://github.com/gitpod-io/gitpod/pull/10048) is merged earlier, but we can also use the feature flag for the error messages.)
2. Select an emoji.
3. Enter a feedback text, or not.
4. Submit feedback.

#### Analytics test:
After sending a feedback in the steps above, go to https://app.segment.com/gitpod/sources/staging_untrusted/debugger and find `feedback_submitted` events.

## Release Notes

```release-note
Users can send feedback from the Dashboard.
```

## Documentation

/werft analytics=segment|TEZnsG4QbLSxLfHfNieLYGF4cDwyFWoe

### CODE REVIEW

**Review by jldec — APPROVED** [review]
/hold pending engineering review
I love it! Excellent UI work (hover states, keyboard interaction)
Q: Will you do separate PR for feedback on error conditions?

File: components/dashboard/src/Analytics.tsx
Comment (jakobhero) [inline components/dashboard/src/Analytics.tsx:35]: hi @laushinka!👋 i changed these two properties to match up with the property names of the `feedback_submitted` call from the website - i hope you don't mind me directly committing them to your baby!🙏🏼

**Review by jakobhero — APPROVED** [review]
lgtm tracking-wise. i noticed that the text in the feedback box was not visible for me unless i highlighted it though - might be due to the dark theme i use...

Screenshot 2022-05-17 at 20 41 31

File: components/dashboard/src/Analytics.tsx
Comment (laushinka) [reply components/dashboard/src/Analytics.tsx:35]: Of course, feel free! Thank you 🤗

Comment (laushinka) [reply components/dashboard/src/Analytics.tsx:35]: Will fix the text issue 🙏🏽

**Comment by laushinka** [comment]
> Will you do separate PR for feedback on error conditions?

Yes, I have a draft PR already, but have more questions. Will ping you.

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/feedback-form/FeedbackModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/feedback-form/FeedbackModal.tsx:20]: **suggestion**: Once we address all the UX comments on the styling of the second step, could we adjust the height of the modal using padding or adjusting the custom textarea height, so that modal has the same height for all steps?

**Review by gtsiolis — COMMENTED** [review]
Left few more notes regarding dark theme layout.

giphy (1)

File: components/dashboard/src/feedback-form/FeedbackModal.tsx
Comment (laushinka) [reply components/dashboard/src/feedback-form/FeedbackModal.tsx:20]: Fixes pushed! Do you mean that the first flow with only the emojis should be the same height as the expanded one?

Comment (laushinka) [reply components/dashboard/src/feedback-form/FeedbackModal.tsx:20]: Done!

Comment (gtsiolis) [reply components/dashboard/src/feedback-form/FeedbackModal.tsx:20]: Looks great!

File: components/dashboard/src/Analytics.tsx
Comment (laushinka) [reply components/dashboard/src/Analytics.tsx]: Weird, I wonder why your commit got lost.. I'll make another one 🙏🏽

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (gtsiolis) [inline components/dashboard/src/feedback-form/FeedbackComponent.tsx:114]: **issue**: Re-posting here from https://github.com/gitpod-io/gitpod/pull/10061#discussion_r875774731 as this is still an issue before 8292fef, and resolving the original comment. Could we use border color for default and focus state as in other text inputs?

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (gtsiolis) [inline components/dashboard/src/feedback-form/FeedbackComponent.tsx:127]: **question(non-blocking)**: Do you think using Gitpod's policy here is better than our policy as described in https://github.com/gitpod-io/gitpod/pull/10061#discussion_r875784777? Asking as we've been using the second pattern elsewhere in the dashboard. Your call.

**Review by gtsiolis — COMMENTED** [review]
Round two, ready! 🥊 

Thanks for the swift changes, @laushinka! 🙏

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (laushinka) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx:113]: Do you know how I can reduce the spacing here? Can't figure it out 🙈

Comment (gtsiolis) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx:113]: > Do you know how I can reduce the spacing here?

Haven't checked, @laushinka! But feel free to leave this as is for now. Additionally, in case it helps, we can also drop the zoom-effect when the emojis are used within the text area.

Comment (laushinka) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx:113]: Done! Used `-space-x`

File: components/dashboard/src/App.tsx
Comment (gtsiolis) [inline components/dashboard/src/App.tsx:376]: **nitpick**: Friendly reminder to remove this route as we no longer need it, right?

Comment (laushinka) [reply components/dashboard/src/App.tsx:376]: Ohh thank you for this!!

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (laushinka) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx:114]: Good point and I actually implement a "Thank you for your feedback" response for the error message implementation (PR incoming after this one is merged). Feel free to send a design idea for the modal!

File: components/dashboard/src/App.tsx
Comment (laushinka) [reply components/dashboard/src/App.tsx]: Done

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (gtsiolis) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx:114]: Maybe something like this could suffice, re-using the copy from the docs feedback widget here.

Modal

See [design specs](https://www.figma.com/file/HknoZvF952qXt3Sk4wLdrq/Low-friction-dashboard-form-to-collect-feedback-about-Gitpod-features-%237925?node-id=1068%3A1523) for grabbing the emoji SVG.

Comment (laushinka) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx:114]: Yes, will do it in a follow-up PR!

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (andrew-farries) [inline components/dashboard/src/feedback-form/FeedbackComponent.tsx:23]: Can we avoid using `any` as the type here?

**Review by andrew-farries — CHANGES_REQUESTED** [review]

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (laushinka) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx]: @andrew-farries Fixed, thank you! 🤗

Comment (andrew-farries) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx]: One way to do this would be to use a [mapped type](https://www.typescriptlang.org/docs/handbook/2/mapped-types.html). Define:

```ts
type EmojiName = "starry" | "happy" | "meh" | "crying";
```

then

```ts
const emojiScore: { [P in EmojiName]: number } = {
        starry: 4,
        happy: 3,
        meh: 2,
        crying: 1,
    };
```

then change each call to `handleClick` to use a string literal:

```ts
onClick={(e) => handleClick("meh")}
```

Comment (laushinka) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx]: @andrew-farries Ooo, thanks! Do you prefer this way or the way I just pushed?

Comment (andrew-farries) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx]: You mean [3d45b5e](https://github.com/gitpod-io/gitpod/pull/10061/commits/3d45b5e68eadbf2dd656df915012282515d433ac)? 

That [fails to build](https://werft.gitpod-dev.com/job/gitpod-build-laushinka-implement-feedback-modal-9928.31) when trying to index into the `emojiScore`.

Comment (laushinka) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx]: Ahh, because of further down. Let me see.

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (andrew-farries) [inline components/dashboard/src/feedback-form/FeedbackComponent.tsx:48]: Is there are a reason why this `` tag has different attributes to the others? If it were the same it would be easier to remove the duplication in this component.

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (gtsiolis) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx:106]: **fyi**: Opened a follow-up issue in https://github.com/gitpod-io/gitpod/issues/10124 to keep track of this.

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (gtsiolis) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx:106]: @laushinka The border colors have been fixed for the light theme! This still needs the textarea :a: background and :b: border color on default and hover state to follow the designs of other text inputs in dark theme. To clarify, I'm reposting some screenshots below. 🌔 

| Feedback | Variable |
|-|-|
| Screenshot 2022-05-19 at 2 00 27 PM | Screenshot 2022-05-19 at 2 00 37 PM |

Comment (laushinka) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx:65]: @gtsiolis I just pushed a fix for the textarea background color, but couldn't figure out the difference in the border colors, so I think I need your help here to tell me where to add which classnames? 🙏🏽

Comment (laushinka) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx:65]: Done, thanks!

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (gtsiolis) [inline components/dashboard/src/feedback-form/FeedbackComponent.tsx:68]: **issue**: Minor color fix, @laushinka sorry if all these variables sounds confusing! Once these components mature enough, we can extract all of these into separate reusable components. 🙏 

```suggestion
                                className="w-full resize-none text-gray-400 dark:text-gray-400 focus:ring-0 focus:border-gray-400 dark:focus:border-gray-400 rounded-md border dark:bg-gray-800 dark:border-gray-500 border-gray-500"
```

| BEFORE | AFTER | PARALLEL FIX |
|-|-|-|
| modal-before | modal-after | modal-fix |

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (gtsiolis) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx:65]: Thanks, @laushinka! Looks good, left one last UX comment below in https://github.com/gitpod-io/gitpod/pull/10061#discussion_r877066717. This is almost at the finish line! 🏁

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (andrew-farries) [inline components/dashboard/src/feedback-form/FeedbackComponent.tsx:38]: I think this can become an `EmojiGroup` component, taking `width` as a prop, rather than a function.

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (andrew-farries) [inline components/dashboard/src/feedback-form/FeedbackComponent.tsx:64]: Use `EmojiGroup` as a component.

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (andrew-farries) [inline components/dashboard/src/feedback-form/FeedbackComponent.tsx:104]: Use `EmojiGroup` as a component.

**Review by andrew-farries — CHANGES_REQUESTED** [review]

File: components/dashboard/src/feedback-form/FeedbackComponent.tsx
Comment (andrew-farries) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx:38]: not necessary, but I always find it odd when a function returns a react element but isn't a component. It doesn't need to be moved to a separate file.

✅ and up to you whether you change it or not 😄

Comment (laushinka) [reply components/dashboard/src/feedback-form/FeedbackComponent.tsx:38]: Ah, interesting thought! I do think in the team there are different views on what best practices we can do in React, and maybe worth having a discussion on 🤩 I will skip this extraction for this one, but keeping my mind open 🤗

**Comment by laushinka** [comment]
Thank you for the reviews, all!

**Pull request merged**
